### PR TITLE
repair issue with deletion of child tasks during sync_project

### PR DESCRIPTION
**Pull request by ghost**
So the assumption on line 273 is that `self.client.tasks.find_all({'project': project_id})` will return *all* task ids for a project, but it appears that it only returns the top level task ids.  Because of this, syncing was deleting the child Task models.   I didn't update the test case as I wasn't positive about this, but I think the test_task_with_parent test mocks the asana API incorrectly in this regard.  

This PR repairs this issue and also adds a debug log to indicate the tasks getting deleted.

The repair creates a list of tasks recursively synced from a call to _sync_task and then uses that as the final list of ids to exclude when deleting tasks.

### CODE REVIEW

**Comment by sbywater** [comment]
Thank you for these commits. I will work this week to test the theory that client.tasks.find_all does not return child tasks.

**Comment by sbywater** [comment]
I tested your theory that find_all does not return child tasks, and did not find that to be true. Here is the code I used to test; please let me know if your experience differs:

```
from djasana.connect import client_connect
client = client_connect()
data = {'workspace': PERSONAL_PROJ_ID, 'owner': MY_ID, 'name': 'Simple Test'}
proj_response = client.projects.create(data)
parent_task = {'name': 'Parent', 'projects': [proj_response['id']], 'workspace': proj_response['workspace']['id']}
parent_response = client.tasks.create(**parent_task)
child_task = {'name': 'Child', 'projects': [proj_response['id']], 'workspace': proj_response['workspace']['id']}
child_response = client.tasks.create(**child_task)
client.tasks.set_parent(child_response['id'], {'parent': parent_response['id']})
tasks = client.tasks.find_all({'project': proj_response['id']})
[t for t in tasks]
```
That returns both the parent and the child task.

**Comment by ghost** [comment]
Hi, thanks for checking that out!  Sorry to go silent.  I have learned after reading your test.

To repeat what I observed, try creating the child task by hand in asana web UI.  If you do this, you may notice that child tasks created via the Asana web UI aren't assigned to the same project as the parent by default?  In your test above, you have assigned the child task to the same project as the parent, so the find_all still works.  This would explain why the real-world child tasks are getting deleted.  

Currently, the sync_from_asana crashes when trying to hook stories to project-less child tasks that have been deleted.  

If you create at child task by hand, and add a comment to it, see if it sync works for you.  Perhaps this is a strange edge case that one of my asana users created by accident.

**Comment by sbywater** [comment]
That is a good explanation of the error you are seeing. I did not anticipate tasks without a project but you are correct, this is something Asana supports so I will endeavour to add support for this.

**Comment by sbywater** [comment]
That is a good explanation of the error you are seeing. I did not anticipate tasks without a project but you are correct, this is something Asana supports so django-asana should support it as well.

File: djasana/management/commands/sync_from_asana.py [hunk 1]
```diff
@@ -271,11 +271,14 @@ def _sync_project_id(self, project_id, workspace, models):
...
-                self._sync_task(task, project, models)
+                ids = self._sync_task(task, project, models)
+                remote_task_ids.extend(ids)
             # Delete local tasks for this project that are no longer in Asana.
-            Task.objects.filter(projects=project).exclude(
-                remote_id__in=remote_task_ids).exclude(remote_id__isnull=True).delete()
+            tasks_to_delete = Task.objects.filter(projects=project).exclude(
+                remote_id__in=remote_task_ids).exclude(remote_id__isnull=True)
+            id_list = list(tasks_to_delete.values_list('remote_id', flat=True))
+            logger.debug("Deleting Tasks No Longer Present " + str(id_list))
```
Comment (sbywater) [inline djasana/management/commands/sync_from_asana.py]: The rest of the module logs to logger and writes to stdout for console feedback; please do that here. 
message = 'Deleting tasks no longer present: {}.'.format(id_list)

File: djasana/management/commands/sync_from_asana.py [hunk 2]
```diff
@@ -307,6 +310,7 @@ def _sync_task(self, task, project, models, skip_subtasks=False):
         For parents and subtasks, this method is called recursively, so skip_subtasks True is
         passed when syncing a parent task from a subtask.
         """
+        was_synced = []
```
Comment (sbywater) [inline djasana/management/commands/sync_from_asana.py]: In my mind, variables is_* and was_* are expected to be boolean, like is_archived. Changing this variable to synced_ids conforms to the style I've used.

**Review by sbywater — CHANGES_REQUESTED** [review]
A few minor requests here. In addition, please write a test that demonstrates this new behavior, and uses a task that is not attached to a project.

File: djasana/management/commands/sync_from_asana.py (hunk 2 above)
Comment (ghost) [reply djasana/management/commands/sync_from_asana.py]: great point

File: djasana/management/commands/sync_from_asana.py (hunk 1 above)
Comment (ghost) [reply djasana/management/commands/sync_from_asana.py]: sounds good, did this update

**Comment by ghost** [comment]
Cleanups are no problem, I committed those.  I am struggling though with setting up the mocks to make this test work correctly.  I have a test I'm working out below.  Perhaps you have some insights? (comments highlight issue)

    def test_child_task_without_project(self):
        parent_task = task()
        parent_copy = parent_task.copy()
        child_task = task(id=2, parent=parent_task)
        child_copy = child_task.copy()
        # the key part of the test is the call to find_all for a project only returns the parent
        self.command.client.tasks.find_all.return_value = [parent_task]
        # and the call to subtask returns the child that didn't appear above
        self.command.client.tasks.subtasks.return_value = [child_task]
        # This is where I am stuck.  I cant seem to get a combination of these return values
        #  that doesn't throw a StopIteration. 
        self.command.client.tasks.find_by_id.side_effect = [parent_task, child_task]

        self.command.handle(interactive=False, project=['Test Project'])
        self.assertEqual(2, Task.objects.count())

**Comment by sbywater** [comment]
A couple things going on here. Every time a mock response gets processed its id is popped, and as a result it cannot be used again, so that's why we send in copies. The way you had it, parent_task had no id  by this point so this code would never find a match:

`if not Task.objects.filter(remote_id=parent_id).exists()`

Also, when processing the child task, we don't want the child task itself to have subtasks. You were mocking the return value so the child task was always a subtask of everything, even itself, and the job would keep processing down that chain until it ran out. 

Here is a version that does what I think you want it to do:

```
def test_child_task_without_project(self):
        parent_task = task(projects=None)
        child_task = task(id=2, projects=None, parent=parent_task.copy())

        self.command.client.tasks.find_all.return_value = [parent_task.copy()]
        self.command.client.tasks.subtasks.side_effect = [[child_task.copy()], []]
        self.command.client.tasks.find_by_id.side_effect = [parent_task.copy(), child_task.copy()]

        self.command.handle(interactive=False, workspace=['Test Workspace'])
        self.assertEqual(2, Task.objects.count())
```

**Comment by ghost** [comment]
Aha!  It was the need to switch the subtask to a side effect that  I wasn’t seeing.  Big thanks.

**Pull request merged**